src/esp/core/managedContainers/ManagedContainerBase.cpp

```diff
@@ -48,53 +48,79 @@ std::string ManagedContainerBase::getRandomObjectHandlePerType(
+ * param)
+ * @tparam The class of the keys in the passed map
+ * @tparam The class of the values in the passed map
+ * @param mapOfHandles The map to search
+ * @param subStr The substring to find within the requested index (key or value)
+ * in the map elements
+ * @param contains Whether @p subStr should be found or excluded in the results
+ * returned
+ * @param sorted Whether the results should be sorted.
+ * @return A vector of strings that either match, or explicitly do not match,
+ * the passed @p subStr in the passed @p mapOfHandles.
+ */
+
+template <int Idx, class T, class U>
+std::vector<std::string> getHandlesBySubStringPerTypeInternal(
+    const std::unordered_map<T, U>& mapOfHandles,
     const std::string& subStr,
     bool contains,
-    bool sorted) const {
+    bool sorted) {
   std::vector<std::string> res;
   // if empty return empty vector
   if (mapOfHandles.size() == 0) {
     return res;
   }
+  res.reserve(mapOfHandles.size());
   // if search string is empty, return all values
   if (subStr.length() == 0) {
-    res.reserve(mapOfHandles.size());
     for (const auto& elem : mapOfHandles) {
-      res.push_back(elem.second);
-    }
-    if (sorted) {
-      std::sort(res.begin(), res.end());
+      res.emplace_back(std::get<Idx>(elem));
     }
-    return res;
-  }
-  // build search criteria
-  std::string strToLookFor = Cr::Utility::String::lowercase(subStr);
-
-  std::size_t strSize = strToLookFor.length();
-
-  for (std::unordered_map<int, std::string>::const_iterator iter =
-           mapOfHandles.begin();
-       iter != mapOfHandles.end(); ++iter) {
-    std::string key = Cr::Utility::String::lowercase(iter->second);
-    // be sure that key is big enough to search in (otherwise find has undefined
-    // behavior)
-    if (key.length() < strSize) {
-      continue;
-    }
-    bool found = (std::string::npos != key.find(strToLookFor));
-    if (found == contains) {
-      // if found and searching for contains, or not found and searching for not
-      // contains
-      res.push_back(iter->second);
+  } else {
+    // build search criteria for reverse map
+    std::string strToLookFor = Cr::Utility::String::lowercase(subStr);
+    std::size_t strSize = strToLookFor.length();
+    for (typename std::unordered_map<T, U>::const_iterator iter =
+             mapOfHandles.begin();
+         iter != mapOfHandles.end(); ++iter) {
+      std::string rawKey = std::get<Idx>(*iter);
+      std::string key = Cr::Utility::String::lowercase(rawKey);
+      // be sure that key is big enough to search in (otherwise find has
+      // undefined behavior)
+      if (key.length() < strSize) {
+        continue;
+      }
+      bool found = (std::string::npos != key.find(strToLookFor));
+      if (found == contains) {
+        // if found and searching for contains, or not found and searching for
+        // not contains
+        res.emplace_back(rawKey);
```

Comment:
```suggestion        
        res.emplace_back(std::move(rawKey));
```

Reply:
plan on combing through all of these (there's quite a few), in this PR.